### PR TITLE
[Fix #50604] Restore compatibility of Active Record Encryption configs with eager loading mode

### LINKED ISSUE #50604: Active Record Encryption configs not ready in eager loading mode
### Description

The introduction of `after_initialize` in Active Record Encryption configuration setup in https://github.com/rails/rails/pull/48530 (also see [that discussion in the same PR](https://github.com/rails/rails/pull/48530/files#r1237624233)) is not fully compatible with the eager loading mode: https://github.com/rails/rails/blob/6b446bee63c401364d193920f3426af0bfe75650/activerecord/lib/active_record/railtie.rb#L385

In eager loading mode, the models from the application are loaded before the `after_initialize` hook has a chance to run, as visible here (hint: you need to scroll in the code snippet): https://github.com/rails/rails/blob/b0b1eb5a1847c1de5616a55c1d7c78f11eb08cd8/railties/lib/rails/application/finisher.rb#L77-L95

As a result, some configurations may not be ready at the time Active Record Encryption's helper `encrypts` needs them during model loading.

For example, the following methods are called in order: `encrypts` ->  `encrypt_attribute` -> `scheme_for` -> `global_previous_schemes_for`:

https://github.com/rails/rails/blob/e0a55b038f7f2f50d1467876558be183be6cedaa/activerecord/lib/active_record/encryption/encryptable_record.rb#L78-L82

If `ActiveRecord::Encryption.config.previous_schemes` is not defined yet, then the encrypted attribute is not properly configured.

#### Steps to reproduce

In this example, a previous encryption scheme is provided globally, as described in https://guides.rubyonrails.org/active_record_encryption.html#global-previous-encryption-schemes:

```ruby
# config/application.rb
config.active_record.encryption.previous = [ { key_provider: MyOldKeyProvider.new } ]
config.eager_load = true

# app/models/article.rb
class Article < ApplicationRecord
  encrypts :title
end
```

#### Expected behavior

The previous encryption scheme should always be taken into account when line `encrypts :title` is reached.

#### Actual behavior

When lazy loading the application, everything is fine and the `Article` model is lazily loaded after the previous encryption scheme is defined by the `after_initialize` hook and the `:title` attribute can benefit from that previous encryption scheme.

When eager loading the application however, the `Article` model is eager loaded before the previous scheme is defined and that config gets ignored.

#### Steps to reproduce

In this example, a custom key provider is configured when declaring the attribute, as described in https://guides.rubyonrails.org/active_record_encryption.html#model-specific-key-providers:

```ruby
# config/application.rb
config.eager_load = true

# app/models/article.rb
class Article < ApplicationRecord
  encrypts :title, key_provider: CustomKeyProvider.new(ActiveRecord::Encryption.config.primary_key)
end
```

#### Expected behavior

The custom key provider should be able to use the `primary_key` configured  when line `encrypts :title` is reached.

#### Actual behavior

When lazy loading the application, everything is fine and the `Article` model is lazily loaded after the `primary_key` is defined by the `after_initialize` hook and the `:title` attribute can be set up normally.

When eager loading the application however, the `Article` model is eager loaded before the `primary_key` is defined by the `after_initialize` hook, and the model fails to load with the following exception:

```
activerecord-7.1.2/lib/active_record/encryption/config.rb:43:in `block (2 levels) in <class:Config>':
  Missing Active Record encryption credential: active_record_encryption.primary_key (ActiveRecord::Encryption::Errors::Configuration)
```

### System configuration

**Rails version**: `7.1.2` / `7.1.3` / `main`

**Ruby version**: `3.2.2` / `3.3.0`

### Possible fixes

1. Get rid of `after_initialize` if not really necessary (see [previous discussions](https://github.com/rails/rails/pull/48530/files#r1237624233))
2. Otherwise, keep `after_initialize` and attempt to patch the eager loading mode (using `config.before_eager_load`?)

### PR DESCRIPTION
**Pull request by maximerety**
### Motivation / Background

Fixes #50604

This Pull Request fixes a source of incompatibility between the loading order of Active Record Encryption configs and the loading of AR models in eager loading mode.

### Detail

This Pull Request removes an occurrence of `after_initialize` introduced in https://github.com/rails/rails/pull/48530, which is not fully compatible with the eager loading mode.

Previous concerns were raised in [a discussion from the same PR](https://github.com/rails/rails/pull/48530/files#r1237624233).

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

**Comment by rafaelfranca** [comment]
This broke a test where the config is set in an intializer. That is because the `ActiveRecord::Encryption.configure` is called outside of the `on_load`. We should move it there.

File: railties/test/application/configuration_test.rb [hunk 1]
```diff
@@ -3819,6 +3826,74 @@ class Post < ActiveRecord::Base
       assert_not_includes ActiveRecord::Base.filter_attributes, :content
     end
 
+    test "ActiveRecord::Encryption.config is ready for encrypted attributes when app is lazy loaded" do
```
Comment (maximerety) [inline railties/test/application/configuration_test.rb:3829]: To clarify, this test is passing before the fix but is added to demonstrate both cases are handled (eager loading and lazy loading).

File: railties/test/application/configuration_test.rb [hunk 2]
```diff
@@ -3819,6 +3826,74 @@ class Post < ActiveRecord::Base
...
+      RUBY
+
+      app "development"
+
+      assert_kind_of ::MyOldKeyProvider, Post.attribute_types["content"].previous_schemes.first.key_provider
+      assert_kind_of ::MyCustomKeyProvider, Post.attribute_types["content"].scheme.key_provider
+      assert_equal "dummy_key", Post.attribute_types["content"].scheme.key_provider.primary_key
+    end
+
+    test "ActiveRecord::Encryption.config is ready for encrypted attributes when app is eager loaded" do
```
Comment (maximerety) [inline railties/test/application/configuration_test.rb:3863]: This is the test which fails before the addition of the fix, with the following error:

```
ActiveRecord::Encryption::Errors::Configuration: Missing Active Record encryption credential: active_record_encryption.primary_key
    /workspaces/rails/activerecord/lib/active_record/encryption/config.rb:43:in `block (2 levels) in <class:Config>'
    /workspaces/rails/tmp/d20240106-41247-luvct9/app/app/models/post.rb:2:in `<class:Post>'
    /workspaces/rails/tmp/d20240106-41247-luvct9/app/app/models/post.rb:1:in `<top (required)>'
    /usr/local/rvm/gems/default/gems/zeitwerk-2.6.12/lib/zeitwerk/kernel.rb:30:in `require'
    /usr/local/rvm/gems/default/gems/zeitwerk-2.6.12/lib/zeitwerk/kernel.rb:30:in `require'
    /usr/local/rvm/gems/default/gems/zeitwerk-2.6.12/lib/zeitwerk/loader/helpers.rb:135:in `const_get'
    /usr/local/rvm/gems/default/gems/zeitwerk-2.6.12/lib/zeitwerk/loader/helpers.rb:135:in `cget'
    /usr/local/rvm/gems/default/gems/zeitwerk-2.6.12/lib/zeitwerk/loader/eager_load.rb:175:in `block in actual_eager_load_dir'
    /usr/local/rvm/gems/default/gems/zeitwerk-2.6.12/lib/zeitwerk/loader/helpers.rb:40:in `block in ls'
    /usr/local/rvm/gems/default/gems/zeitwerk-2.6.12/lib/zeitwerk/loader/helpers.rb:25:in `each'
    /usr/local/rvm/gems/default/gems/zeitwerk-2.6.12/lib/zeitwerk/loader/helpers.rb:25:in `ls'
    /usr/local/rvm/gems/default/gems/zeitwerk-2.6.12/lib/zeitwerk/loader/eager_load.rb:170:in `actual_eager_load_dir'
    /usr/local/rvm/gems/default/gems/zeitwerk-2.6.12/lib/zeitwerk/loader/eager_load.rb:17:in `block (2 levels) in eager_load'
    /usr/local/rvm/gems/default/gems/zeitwerk-2.6.12/lib/zeitwerk/loader/eager_load.rb:16:in `each'
    /usr/local/rvm/gems/default/gems/zeitwerk-2.6.12/lib/zeitwerk/loader/eager_load.rb:16:in `block in eager_load'
    /usr/local/rvm/gems/default/gems/zeitwerk-2.6.12/lib/zeitwerk/loader/eager_load.rb:10:in `synchronize'
    /usr/local/rvm/gems/default/gems/zeitwerk-2.6.12/lib/zeitwerk/loader/eager_load.rb:10:in `eager_load'
    /usr/local/rvm/gems/default/gems/zeitwerk-2.6.12/lib/zeitwerk/loader.rb:379:in `block in eager_load_all'
    /usr/local/rvm/gems/default/gems/zeitwerk-2.6.12/lib/zeitwerk/loader.rb:377:in `each'
    /usr/local/rvm/gems/default/gems/zeitwerk-2.6.12/lib/zeitwerk/loader.rb:377:in `eager_load_all'
    /workspaces/rails/railties/lib/rails/application/finisher.rb:80:in `block in <module:Finisher>'
    /workspaces/rails/railties/lib/rails/initializable.rb:32:in `instance_exec'
    /workspaces/rails/railties/lib/rails/initializable.rb:32:in `run'
    /workspaces/rails/railties/lib/rails/initializable.rb:61:in `block in run_initializers'
    /usr/local/lib/ruby/3.2.0/tsort.rb:228:in `block in tsort_each'
    /usr/local/lib/ruby/3.2.0/tsort.rb:350:in `block (2 levels) in each_strongly_connected_component'
    /usr/local/lib/ruby/3.2.0/tsort.rb:431:in `each_strongly_connected_component_from'
    /usr/local/lib/ruby/3.2.0/tsort.rb:349:in `block in each_strongly_connected_component'
    /usr/local/lib/ruby/3.2.0/tsort.rb:347:in `each'
    /usr/local/lib/ruby/3.2.0/tsort.rb:347:in `call'
    /usr/local/lib/ruby/3.2.0/tsort.rb:347:in `each_strongly_connected_component'
    /usr/local/lib/ruby/3.2.0/tsort.rb:226:in `tsort_each'
    /usr/local/lib/ruby/3.2.0/tsort.rb:205:in `tsort_each'
    /workspaces/rails/railties/lib/rails/initializable.rb:60:in `run_initializers'
    /workspaces/rails/railties/lib/rails/application.rb:423:in `initialize!'
    /workspaces/rails/tmp/d20240106-41247-luvct9/app/config/environment.rb:5:in `<top (required)>'
    /usr/local/rvm/gems/default/gems/zeitwerk-2.6.12/lib/zeitwerk/kernel.rb:38:in `require'
    /usr/local/rvm/gems/default/gems/zeitwerk-2.6.12/lib/zeitwerk/kernel.rb:38:in `require'
    /workspaces/rails/railties/test/application/configuration_test.rb:59:in `app'
    /workspaces/rails/railties/test/application/configuration_test.rb:3890:in `block in <class:ConfigurationTest>'
```

**Review by maximerety — COMMENTED** [review]
> This broke a test where the config is set in an intializer. That is because the `ActiveRecord::Encryption.configure` is called outside of the `on_load`. We should move it there.

@rafaelfranca Thanks for the hint.

I had a hard time adding new tests to demonstrate the issue because, to be honest, there are a lot of possible situations that could still not work as expected.

For example, people may reference `ActiveRecord::Base` in initializers (this is what is done in the new tests here, and others) and if you do this before setting the desired `Rails.application.config.active_record.encryption` properties, they will be ignored when configuring `ActiveRecord::Encryption`. That's why I'm regrouping all configs in a single initializer (`config/initializers/active_record.rb`) in the tests and declare them in the proper order.

This is the exact problem discussed initially: https://github.com/rails/rails/pull/48530/files#r1237624233.

But this is probably the price we have to pay to ensure that configurations are ready early enough in eager loading mode.

Finally, I'd like to propose a backport of this fix to `7-1-stable` as I encountered it while migrating a project from 7.0 to 7.1. Would that be acceptable to you?

**Review by jorgemanrubia — APPROVED** [review]
Good fix 👍 , thanks @maximerety.

I think this case cause trouble when referencing models from initializers, but @rafaelfranca made a good point here  about that https://github.com/rails/rails/pull/48530#discussion_r1242393461. Good to merge from my side 👍 .

Also, thanks for the fantastic stream of improvements on Active Record Encryption @maximerety 🙏.

**Comment by maximerety** [comment]
Thanks @jorgemanrubia for your kind words 🙇 

@rafaelfranca Let me know if you need anything else from me for this PR to merged and backported to `7-1-stable` 🙏

**Comment by maximerety** [comment]
@rafaelfranca I've rebased and squashed the 2 commits (which I forgot to do initially).

**Comment by maximerety** [comment]
@jorgemanrubia Too bad I did not manage to have this merged in Rails 7.1.3, I think I'll have to be little more patient 😉 Do you know if I can do something else for this patch to land on `main` / `7-1-stable` so that it'll be ready for a next patch version when that times come? Thanks.

**Comment by trevorturk** [comment]
@maximerety you might consider pinging the Rails Discord to see if you can get a review?

**Comment by jorgemanrubia** [comment]
@eileencodes could we merge this one please?

**Comment by eileencodes** [comment]
Can someone do the backport PR? It doesn't cherry-pick cleanly and I don't have the bandwidth to fight conflicts right now. Thanks!

**Comment by maximerety** [comment]
Thanks @eileencodes, here is the backport PR to 7-1-stable: https://github.com/rails/rails/pull/51061.

**Comment by tenpaiyomi** [comment]
@jorgemanrubia @eileencodes I don't suppose either of you have an idea of when this will actually make it into a release, do you?  I saw that the backport to 7-1-stable, #51061, got merged in about 4 months ago, and there seems to have been 3 separate 7.1.x releases since then but none of them actually include this fix which as left me stuck having to pin my rails gem version to the `7-1-stable` branch instead of updating to the latest gem release.
 
 So far it seems like it is on track to release with 7.2.x (it exists in the v7.2.0.beta2) but my thought with the backport to the `7-1-stable` branch PR was that it would be included in 7.1.x releases going forward as well.

**Comment by zzak** [comment]
@tenpaiyomi The last bugfix release was 7.1.3, before the backport was merged. So whenever 7.1.4 is out, that should include this. Thanks for your patience.

**Pull request merged**